bgpd/bgp_fsm.c

```diff
@@ -139,8 +139,25 @@ peer_xfer_conn(struct peer *from_peer)
   peer->fd = from_peer->fd;
   from_peer->fd = fd;
   stream_reset(peer->ibuf);
-  stream_fifo_clean(peer->obuf);
-  stream_fifo_clean(from_peer->obuf);
+
+  // At this point in time, it is possible that there are packets pending on
+  // from_peer->obuf. These need to be transferred to the new peer struct.
+  pthread_mutex_lock (&peer->obuf_mtx);
+  pthread_mutex_lock (&from_peer->obuf_mtx);
+  {
+    // wipe new peer's packet queue
+    stream_fifo_clean (peer->obuf);
+
+    // copy each packet from old peer's queue to new peer's queue
+    struct stream *s;
+    for (s = from_peer->obuf->head; s; s = s->next)
+      stream_fifo_push (peer->obuf, stream_dup (s));
```

Comment:
Would something like this be better?

while (from_peer->obuf->head)
    stream_fifo_push (peer->obuf, stream_fifo_pop(from_peer->obuf);

Avoiding the Stream_dup? and the stream_fifo_clean immediately below?